### PR TITLE
🗻 autograd: dispersive models

### PR DESCRIPTION
**Pull request by tylerflex**
The eventual goal is to do topology optimization with custom dispersive models. This PR will head in that direction.

Breaking it into steps:

- [x] `PoleResidue` support
    - [x] `.eps_inf` support
    - [x] `.poles` support
- [x] support for `CustomPoleResiude`
    - [x] demo notebook, antenna invdes?

## For another time...
- [ ] support for other dispersive models (Lorentz, etc)
- [ ] support for `CustomLoretnz` etc.

### CODE REVIEW

**Comment by tylerflex** [comment]
This is basically ready to go. I want to do a bit more testing on the antenna [notebook](https://github.com/flexcompute/tidy3d-notebooks/pull/124) first, though. just to ensure that the gradients are working properly.

File: tidy3d/components/autograd/derivative_utils.py [hunk 1]
```diff
@@ -70,6 +70,12 @@ class DerivativeInfo(Tidy3dBaseModel):
         description="Bounds corresponding to the structure, used in ``Medium`` calculations.",
     )
 
+    frequency: float = pd.Field(
+        ...,
+        title="Frequency of adjoint simulation",
+        description="Frequency at which the adjoint gradient is computed.",
+    )
+
```
Comment (yaugenst-flex) [inline tidy3d/components/autograd/derivative_utils.py:78]: Does this imply that we'll have a separate `DerivativeInfo` for each frequency?

File: tidy3d/components/base.py [hunk 2]
```diff
@@ -212,7 +212,7 @@ def _special_characters_not_in_name(cls, values):
 
     def copy(self, deep: bool = True, **kwargs) -> Tidy3dBaseModel:
         """Copy a Tidy3dBaseModel.  With ``deep=True`` as default."""
-        kwargs.update(deep=deep)
+        # kwargs.update(deep=deep)
```
Comment (yaugenst-flex) [inline tidy3d/components/base.py]: intentional?

File: tidy3d/components/medium.py [hunk 3]
```diff
@@ -1321,6 +1359,60 @@ def _not_loaded(field):
...
             return any(len(subfield) == 0 for subfield in [field.points, field.cells, field.values])
 
+    def _derivative_field_cmp(
+        self,
+        E_der_map: ElectromagneticFieldDataset,
+        eps_data: PermittivityDataset,
+        dim: str,
+    ) -> np.ndarray:
+        coords_interp = {key: val for key, val in eps_data.coords.items() if len(val) > 1}
+        dims_sum = {dim for dim in eps_data.coords.keys() if dim not in coords_interp}
```
Comment (yaugenst-flex) [inline tidy3d/components/medium.py:1369]: Writing mostly so I can comment on dim sum 😂 Maybe style: `dims_sum = set(eps_data.coords.keys()) - set(coords_interp)` ... the comprehension might be more accessible though so idk.

File: tidy3d/components/medium.py [hunk 4]
```diff
@@ -3092,6 +3184,49 @@ def loss_upper_bound(self) -> float:
...
+        dJ_deps = self.derivative_eps_complex_volume(
+            E_der_map=derivative_info.E_der_map, bounds=derivative_info.bounds
+        )
+
+        dJ_deps = complex(dJ_deps)
+
+        # TODO: fix for multi-frequency
+        frequency = derivative_info.frequency
+        poles_complex = [(complex(a), complex(c)) for a, c in self.poles]
+        poles_complex = np.stack(poles_complex, axis=0)
```
Comment (yaugenst-flex) [inline tidy3d/components/medium.py:3200]: Maybe just `poles_complex = np.array([(complex(a), complex(c)) for a, c in self.poles])`? `np.stack` is identical to `np.array` if `axis=0`.

File: tidy3d/components/medium.py [hunk 5]
```diff
@@ -3092,6 +3184,49 @@ def loss_upper_bound(self) -> float:
...
+            # ignore warnings about holmorphic grad being passed a non-complex input (poles)
+            warnings.simplefilter("ignore")
+            deps_deps_inf, deps_dpoles = grad_eps_model(
+                complex(self.eps_inf), poles_complex, complex(frequency)
+            )
+
+        # multiply with partial dJ/deps to give full gradients
+
+        dJ_deps_inf = dJ_deps * deps_deps_inf
+        dJ_dpoles = [(dJ_deps * a, dJ_deps * c) for (a, c) in deps_dpoles]
```
Comment (yaugenst-flex) [inline tidy3d/components/medium.py]: Nitpick, but the explicit tuple isn't necessary, i.e. `for (a, c) in deps_dpoles` --> `for a, c in deps_dpoles`

File: tidy3d/components/medium.py [hunk 6]
```diff
@@ -3092,6 +3184,49 @@ def loss_upper_bound(self) -> float:
...
+
+        # get vjps w.r.t. permittivity and conductivity of the bulk
+        derivative_map = {}
+        for field_path in derivative_info.paths:
+            field_name, *rest = field_path
+
+            if field_name == "eps_inf":
+                derivative_map[field_path] = float(np.real(dJ_deps_inf))
+
+            if field_name == "poles":
```
Comment (yaugenst-flex) [inline tidy3d/components/medium.py]: Maybe `elif`? Clearer and more efficient.

File: tidy3d/components/medium.py [hunk 7]
```diff
@@ -3335,6 +3470,75 @@ def _sel_custom_data_inside(self, bounds: Bound):
...
+        """Compute adjoint derivatives for each of the ``fields`` given the multiplied E and D."""
+
+        dJ_deps = 0.0
+        for dim in "xyz":
+            dJ_deps += self._derivative_field_cmp(
+                E_der_map=derivative_info.E_der_map, eps_data=self.eps_inf, dim=dim
+            )
+
+        # import pdb; pdb.set_trace()
+        # dJ_deps = np.conj(dJ_deps)
```
Comment (yaugenst-flex) [inline tidy3d/components/medium.py]: Remove?

File: tidy3d/components/medium.py [hunk 8]
```diff
@@ -3335,6 +3470,75 @@ def _sel_custom_data_inside(self, bounds: Bound):
...
+            dJ_dpoles.append((dJ_da, dJ_dc))
+
+        derivative_map = {}
+        for field_path in derivative_info.paths:
+            field_name, *rest = field_path
+
+            if field_name == "eps_inf":
+                derivative_map[field_path] = np.real(dJ_deps_inf)
+
+            if field_name == "poles":
```
Comment (yaugenst-flex) [inline tidy3d/components/medium.py]: `elif`?

File: tidy3d/plugins/autograd/README.md [hunk 9]
```diff
@@ -152,6 +158,7 @@ Next on our roadmap (targeting 2.8 and 2.9, summer 2024) is to support:
 - `PoleResidue` and other dispersive models.
 - custom (spatially-dependent) dispersive models, allowing topology optimization with metals.
 
+- `ComplexPolySlab`
```
Comment (yaugenst-flex) [inline tidy3d/plugins/autograd/README.md]: Is this related?

**Review by yaugenst-flex — COMMENTED** [review]
Super nice to have this. I guess there are some bits and pieces that might needs some more digging into in the future, like those scaling factors 😄

File: tidy3d/components/autograd/derivative_utils.py (hunk 1 above)
Comment (tylerflex) [reply tidy3d/components/autograd/derivative_utils.py:78]: No, in the multi-frequency branch, I will change this to a list of frequencies. for now it's just single, though.

File: tidy3d/components/base.py (hunk 2 above)
Comment (tylerflex) [reply tidy3d/components/base.py]: no! was testing something and forgot to change it back. sorry about that.

File: tidy3d/components/medium.py (hunk 3 above)
Comment (tylerflex) [reply tidy3d/components/medium.py:1369]: haha

File: tidy3d/components/medium.py (hunk 5 above)
Comment (tylerflex) [reply tidy3d/components/medium.py]: I can change it, but curious why it wouldn't be caught by ruff?

File: tidy3d/components/medium.py (hunk 4 above)
Comment (tylerflex) [reply tidy3d/components/medium.py:3200]: hm, sort of like the explicit axis because it helps me keep track of the fact that the poles are getting stacked along the first axis. but maybe array is fine.

File: tidy3d/components/medium.py (hunk 5 above)
Comment (yaugenst-flex) [reply tidy3d/components/medium.py]: Hmm yeah interesting. Parentheses are removed in tuple unpacking in for loops, but are left unchanged in  list comprehensions. The choice seems somewhat arbitrary...

File: tidy3d/plugins/autograd/README.md (hunk 9 above)
Comment (tylerflex) [reply tidy3d/plugins/autograd/README.md]: no, should be removed. fixed

**Review by yaugenst-flex — APPROVED** [review]
